### PR TITLE
PIM 1015

### PR DESCRIPTION
**Pull request by pavlovdog**
- Websocket support on `/:version/rpc`

### CODE REVIEW

**Review by mouseless-eth — APPROVED** [review]

**Comment by plusminushalf** [comment]
rest lgtm! 🚀

File: src/rpc/server.ts [hunk 1]
```diff
@@ -113,16 +122,32 @@ export class Server {
...
         })
 
-        this.fastify.post("/rpc", this.rpc.bind(this))
-        this.fastify.post("/:version/rpc", this.rpc.bind(this))
-        this.fastify.post("/", this.rpc.bind(this))
+        this.fastify.post("/rpc", this.rpcHttp.bind(this))
+        this.fastify.post("/:version/rpc", this.rpcHttp.bind(this))
+        this.fastify.post("/", this.rpcHttp.bind(this))
+
+        this.fastify.register(async () => {
```
Comment (nikmel2803) [inline src/rpc/server.ts:136]: can you clarify why do we need this wrapper here pls? as i understand it's something like fastify plugin? but why is it needed here?

File: src/utils/reply-middleware.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import { FastifyReply } from "fastify"
+import * as WebSocket from "ws"
+
+class ReplyMiddleware {
```
Comment (nikmel2803) [inline src/utils/reply-middleware.ts:4]: little bit confused by a name 'middleware', maybe smth like RpcReply would be better?

File: src/rpc/server.ts [hunk 3]
```diff
@@ -113,16 +122,32 @@ export class Server {
...
+            this.fastify.route({
+                method: "GET",
+                url: "/:version/rpc",
+                handler: async (request, reply) => {
+                    await reply.status(500).send("Not implemented")
```
[the remark this answers is not in the document]
Comment (nikmel2803) [reply src/rpc/server.ts:141]: and i believe we should return 404 error, not 5xx

File: src/rpc/server.ts (hunk 1 above)
Comment (pavlovdog) [reply src/rpc/server.ts:136]: you mean why do we need `fastify.register`? As I see `.get` shortcut is not implemented for ws handlers

File: src/utils/reply-middleware.ts (hunk 2 above)
Comment (pavlovdog) [reply src/utils/reply-middleware.ts:4]: Agree

File: src/rpc/server.ts [hunk 4]
```diff
@@ -89,6 +92,12 @@ export class Server {
             disableRequestLogging: true
         })
 
+        this.fastify.register(websocket, {
+            options: {
+                maxPayload: 1048576 // maximum allowed messages size is 1 MiB
```
[the remark this answers is not in the document]
Comment (pavlovdog) [reply src/rpc/server.ts]: Done

File: src/rpc/server.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (pavlovdog) [reply src/rpc/server.ts]: Done

File: src/rpc/server.ts (hunk 1 above)
Comment (pavlovdog) [reply src/rpc/server.ts:137]: Done

**Review by nikmel2803 — APPROVED** [review]

**Pull request merged**